### DISCUSSION TITLE
Compiling IPOPT with MKL Pardiso on Windows

### DISCUSSION THREAD

**Discussion by ottmar-zittlau** [coin-or/Ipopt#602] [Q&A] [answered]
Hi,

I managed to build ipopt on Windows using Intel MKL Pardiso. 
However, when I run `make test` the tests fail with segfaults:
```
Testing C++ Example...
./run_unitTests: line 21: 10478 Segmentation fault      $@ > tmpfile 2>&1
0
 ---- 8< ---- Start of test program output ---- 8< ----
 ---- 8< ----  End of test program output  ---- 8< ----

    ******** Test FAILED! ********
Test program existed with nonzero status.
Testing C Example...
./run_unitTests: line 21: 10480 Segmentation fault      $@ > tmpfile 2>&1
0
 ---- 8< ---- Start of test program output ---- 8< ----
 ---- 8< ----  End of test program output  ---- 8< ----

    ******** Test FAILED! ********
Test program existed with nonzero status.
Skip testing Fortran Example (hs071_f not available)
Skip testing Java Example (Java interface not build)
Testing sIpopt Example parametric_cpp...
./run_unitTests: line 21: 10482 Segmentation fault      $@ > tmpfile 2>&1
0
 ---- 8< ---- Start of test program output ---- 8< ----
 ---- 8< ----  End of test program output  ---- 8< ----

    ******** Test FAILED! ********
Test program existed with nonzero status.
Testing sIpopt Example redhess_cpp...
./run_unitTests: line 21: 10484 Segmentation fault      $@ > tmpfile 2>&1
0
 ---- 8< ---- Start of test program output ---- 8< ----
 ---- 8< ----  End of test program output  ---- 8< ----

    ******** Test FAILED! ********
Test program existed with nonzero status.
Testing EmptyNLP Example...
./run_unitTests: line 21: 10486 Segmentation fault      $@ > tmpfile 2>&1
0
 ---- 8< ---- Start of test program output ---- 8< ----

*** Solve for 0 variables, feasible constraint, feasible bounds
 ---- 8< ----  End of test program output  ---- 8< ----

    ******** Test FAILED! ********
Test program existed with nonzero status.
Testing GetCurr Example...
./run_unitTests: line 21: 10488 Segmentation fault      $@ > tmpfile 2>&1
0
 ---- 8< ---- Start of test program output ---- 8< ----
 ---- 8< ----  End of test program output  ---- 8< ----

    ******** Test FAILED! ********

```

I took the following steps to build ipopt using VS Build Tools 2022.

1. Installed MSYS2 and the packages for it as described in the documentation (https://coin-or.github.io/Ipopt/INSTALL.html).
2. Downloaded and installed Intel MKL.
3. Created a batch file similar to the one given in the documentation:
```
set MSYS2_PATH_TYPE=inherit
call "C:\Libraries\Intel_MKL\setvars.bat"
C:\msys64\msys2.exe
exit 0
```
4. Run the batch file
5. Changed the MKLROOT environment variable to the format used in MSYS2:
`export MKLROOT=$(cygpath -u ${MKLROOT})`
6. Create the build directory and changed into it
```
cd /C/Tools/Ipopt
mkdir build
cd build
```
8. Run the configuration script
```
`../configure --with-lapack-lflags="-L${MKLROOT}/lib/intel64 -lmkl_intel_ilp64_dll -lmkl_sequential_dll -lmkl_core_dll"`
```
I've found the necessary linker options using the Intel Link Advisor tool (https://www.intel.com/content/www/us/en/developer/tools/oneapi/onemkl-link-line-advisor.html).

9. `make` runs through successfully.
10. `make test` fails with the error message above.

Is there anything I can do to fix this?

Thanks in advance and best regards
oz

**Comment by svigerske** [comment, marked as answer]
That's the wrong integer type. You want `mkl_intel_lp64_dll` instead of `mkl_intel_ilp64_dll`  (unless you also build Ipopt for 64-bit integers).

That would then also correspond to the libs that configure tries if you don't specify the lapack flags explicitly (but that requires the MKL lib path to be in the standard library search path).

**Reply by ottmar-zittlau** [reply]
Excellent!

I tried both the 32-bit
```
 ../configure --with-lapack-lflags="-L${MKLROOT}/lib/intel64 -lmkl_intel_lp64_dll -lmkl_sequential_dll -lmkl_core_dll" 
 ``` 
 and the 64-bit version
```
../configure --with-intsize=64 --with-lapack-lflags="-L${MKLROOT}/lib/intel64 -lmkl_intel_ilp64_dll -lmkl_sequential_dll -lmkl_core_dll"
```
and both worked.

And then I also tried to run configure
```
 ../configure
```
without explicitly specifying anything, which also worked.

Thanks for the quick reply!

**Comment by ottmar-zittlau** [comment]
Now that I've built the ipopt-library, I don't manage to link to it.
I've created a small cmake-project that doesn't use ipopt, yet.
main.cpp:
```
#include <iostream>

int main(int, char**) {
    std::cout << "Hello, world!\n";
}
```

CMakeLists.txt:
```
cmake_minimum_required(VERSION 3.0.0)
project(IpoptTest VERSION 0.1.0)

set(IPOPT_DIR "C:/Libs/Ipopt/bin")
find_library(IPOPT NAMES "ipopt" "Ipopt" "ipopt-3" "ipopt-3.dll" PATHS "${IPOPT_DIR}")
message("${IPOPT}")

add_executable(IpoptTest main.cpp)
target_include_directories(IpoptTest PRIVATE "${IPOPT_DIR}/include")
target_link_libraries(IpoptTest PRIVATE "${IPOPT_DIR}/bin/ipopt-3.dll")

```

Find-library doesn't find anything (the variable IPOPT contains "IPOPT_NOTFOUND") and when I build the project, I receive the error
```
[build] C:\Libs\Ipopt\bin\ipopt-3.dll : fatal error LNK1107: Ungültige oder beschädigte Datei: Lesen bei 0x308 nicht möglich.
```

I also tried find-library with the /lib and /bin path.

What am I doing wrong?

Thanks in advance and best regards,
oz

**Reply by svigerske** [reply]
You need to link the `.lib` file, not the `.dll`.

**Reply by ottmar-zittlau** [reply]
Thanks again - that works. 

Now I tried to implement the example given here: https://coin-or.github.io/Ipopt/INTERFACES.html#INTERFACE_CPP
When I run the code, it crashes - apparently when the first option is set
 ```
 app->Options()->SetNumericValue("tol", 3.82e-6);
 ```
 I suspect that the `IpoptApplicationFactory` somehow fails. Is there a way to debug this?
 
 Before running the program I copied `ipopt-3.dll` and `mkl_sequential.2.dll` next to it. Do I need to copy the other mkl-dlls, too?
